### PR TITLE
[ConstraintElim] Extend `checkOrAndOpImpliedByOther` to handle and/or expr trees.

### LINKED ISSUE #117107: signed comparison case missed by constraint elimination
starting with this:
```c
extern void side_effect(void);

void f(int v0, int v1, int v2) {
  if (v0 < v1) return;
  if (v1 < v2) return;
  if (v0 < v2) side_effect();
  return;
}
```

here's what constraint elimination sees:
```llvm
define void @f(i32 noundef %v0, i32 noundef %v1, i32 noundef %v2) {
entry:
  %cmp = icmp sge i32 %v0, %v1
  %cmp1 = icmp sge i32 %v1, %v2
  %or.cond.not11 = and i1 %cmp, %cmp1
  %cmp4 = icmp slt i32 %v0, %v2
  %or.cond10 = and i1 %cmp4, %or.cond.not11
  br i1 %or.cond10, label %if.then5, label %return

if.then5:                                         ; preds = %entry
  tail call void @side_effect() #2
  br label %return

return:                                           ; preds = %if.then5, %entry
  ret void
}
```

this should all optimize away: https://alive2.llvm.org/ce/z/EPLL_9

but it doesn't: https://gcc.godbolt.org/z/enxd5dqxr

the debug output looks sensible, the relations that we want are getting picked up by the constraint system, but I'm not deeply enough into the code to see where we're going wrong
```
Processing condition to simplify:   %cmp = icmp sge i32 %v0, %v1
Checking   %cmp = icmp sge i32 %v0, %v1
---
%v0 + -1 * %v1 <= -1
sat
---
-1 * %v0 + %v1 <= 0
sat
Adding 'icmp sge i32 %v1, %v2'
  constraint: -1 * %v1 + %v2 <= 0

Checking   %cmp = icmp sge i32 %v0, %v1
---
-1 * %v1 + %v2 <= 0
%v0 + -1 * %v1 <= -1
sat
---
-1 * %v1 + %v2 <= 0
-1 * %v0 + %v1 <= 0
sat
Processing condition to simplify:   %cmp1 = icmp sge i32 %v1, %v2
Checking   %cmp1 = icmp sge i32 %v1, %v2
---
%v1 + -1 * %v2 <= -1
sat
---
-1 * %v1 + %v2 <= 0
sat
Adding 'icmp sge i32 %v0, %v1'
  constraint: -1 * %v0 + %v1 <= 0

Checking   %cmp1 = icmp sge i32 %v1, %v2
---
-1 * %v0 + %v1 <= 0
%v1 + -1 * %v2 <= -1
sat
---
-1 * %v0 + %v1 <= 0
-1 * %v1 + %v2 <= 0
sat
Processing condition to simplify:   %cmp4 = icmp slt i32 %v0, %v2
Checking   %cmp4 = icmp slt i32 %v0, %v2
---
-1 * %v0 + %v2 <= 0
sat
---
%v0 + -1 * %v2 <= -1
sat
Processing fact to add to the system: icmp slt i32 %v0, %v2
Adding 'icmp slt i32 %v0, %v2'
  constraint: %v0 + -1 * %v2 <= -1

---
%v0 + -1 * %v2 <= -1
%v0 <= -1
sat
Top of stack : 1 2
CB: 1 2
Processing fact to add to the system: icmp sge i32 %v0, %v1
Adding 'icmp sge i32 %v0, %v1'
  constraint: -1 * %v0 + %v1 <= 0

---
%v0 + -1 * %v2 <= -1
-1 * %v0 + %v1 <= 0
%v1 <= -1
sat
Top of stack : 1 2
CB: 1 2
Processing fact to add to the system: icmp sge i32 %v1, %v2
Adding 'icmp sge i32 %v1, %v2'
  constraint: -1 * %v1 + %v2 <= 0

---
%v0 + -1 * %v2 <= -1
-1 * %v0 + %v1 <= 0
-1 * %v1 + %v2 <= 0
%v2 <= -1
unsat
Adding 'icmp uge i32 %v1, %v2'
  constraint: -1 * %v1 + %v2 <= 0
```
cc @fhahn

### PR DESCRIPTION
**Pull request by dtcxzyw**
This patch extends `checkOrAndOpImpliedByOther` to handle and/or trees.
Limitation: At least one of the operands of root and/or instruction should be an icmp. That is, this patch doesn't support expressions like `(cmp1 & cmp2) & (cmp3 & cmp4)`. To handle this kind of pattern, please see https://github.com/llvm/llvm-project/pull/117118.

Closes https://github.com/llvm/llvm-project/issues/117107.
Compile-time impact: http://llvm-compile-time-tracker.com/compare.php?from=69cc3f096ccbdef526bbd5a065a25c95122e87ee&to=919416d2c4c71e3b9fe533af2c168a36c7893be5&stat=instructions%3Au

### CODE REVIEW

File: llvm/test/Transforms/ConstraintElimination/unreachable-bb.ll [hunk 1]
```diff
@@ -0,0 +1,36 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 5
+; RUN: opt -p constraint-elimination -S %s | FileCheck %s
+
+define void @f(i32 noundef %v0, i32 noundef %v1, i32 noundef %v2) {
```
Comment (fhahn) [inline llvm/test/Transforms/ConstraintElimination/unreachable-bb.ll:4]: could you add variants of the test with different permutations of the AND's and add a version with OR's as well?

File: llvm/test/Transforms/ConstraintElimination/unreachable-bb.ll
Comment (nikic) [inline llvm/test/Transforms/ConstraintElimination/unreachable-bb.ll:1]: unreachable-bb.ll is not really what is pertinent here, this should be part of and-implied-by-operands.ll.

File: llvm/lib/Transforms/Scalar/ConstraintElimination.cpp [hunk 2]
```diff
@@ -1511,22 +1508,40 @@ static bool checkOrAndOpImpliedByOther(
...
+  bool IsOr = match(JoinOp, m_LogicalOr());
+  SmallVector<Value *, 4> Worklist({JoinOp->getOperand(OtherOpIdx)});
+  while (!Worklist.empty()) {
+    Value *Val = Worklist.pop_back_val();
+    Value *LHS, *RHS;
+    ICmpInst::Predicate Pred;
+    if (match(Val, m_ICmp(Pred, m_Value(LHS), m_Value(RHS)))) {
+      if (IsOr)
+        Pred = CmpInst::getInversePredicate(Pred);
+      Info.addFact(Pred, LHS, RHS, CB.NumIn, CB.NumOut, DFSInStack);
```
Comment (fhahn) [inline llvm/lib/Transforms/Scalar/ConstraintElimination.cpp:1531]:
```suggestion
      // Optimistically add fact from the other compares in the AND/OR.
      Info.addFact(Pred, LHS, RHS, CB.NumIn, CB.NumOut, DFSInStack);
```

**Review by nikic — COMMENTED** [review]
The implementation looks fine to me, but the test coverage here is really light.

Do we have a negative test for incorrect and/or mixing?

File: llvm/lib/Transforms/Scalar/ConstraintElimination.cpp [hunk 3]
```diff
@@ -1511,22 +1508,40 @@ static bool checkOrAndOpImpliedByOther(
...
+    while (OldSize < DFSInStack.size()) {
+      StackEntry E = DFSInStack.back();
+      removeEntryFromStack(E, Info, ReproducerModule, ReproducerCondStack,
+                           DFSInStack);
+    }
+  });
+  // For OR, check if the negated condition implies CmpToCheck.
+  bool IsOr = match(JoinOp, m_LogicalOr());
+  SmallVector<Value *, 4> Worklist({JoinOp->getOperand(OtherOpIdx)});
+  while (!Worklist.empty()) {
```
Comment (fhahn) [inline llvm/lib/Transforms/Scalar/ConstraintElimination.cpp:1524]: Would be good to document the loop here

File: llvm/lib/Transforms/Scalar/ConstraintElimination.cpp [hunk 4]
```diff
@@ -1511,22 +1508,41 @@ static bool checkOrAndOpImpliedByOther(
...
+      Info.addFact(Pred, LHS, RHS, CB.NumIn, CB.NumOut, DFSInStack);
+      continue;
+    }
+    if (IsOr ? match(Val, m_LogicalOr(m_Value(LHS), m_Value(RHS)))
+             : match(Val, m_LogicalAnd(m_Value(LHS), m_Value(RHS)))) {
+      Worklist.push_back(LHS);
+      Worklist.push_back(RHS);
+      continue;
+    }
+    return false;
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/Scalar/ConstraintElimination.cpp:1541]: It is safe to ignore unknown leaf nodes here. But I need some positive tests.
```suggestion
```

**Comment by regehr** [comment]
thanks for doing this compile-time study! that would have been my main worry. did you happen to check if this makes constraint elimination fire more in practice? do we have benchmarks (with UBSan turned on for example) that contain extensive opportunities for constraint elimination?

of course it's not a substitute for testing, but my ongoing search for missed optimizations is also a bug hunt, though it's probably the case that purpose-built fuzzers are better at that job.

**Comment by nikic** [comment]
@regehr You can see the impact on real-world code at https://github.com/dtcxzyw/llvm-opt-benchmark/pull/1735.

File: llvm/lib/Transforms/Scalar/ConstraintElimination.cpp (hunk 4 above)
Comment (nikic) [reply llvm/lib/Transforms/Scalar/ConstraintElimination.cpp:1541]: Do you plan to adjust this in this PR or a followup?

Comment (dtcxzyw) [reply llvm/lib/Transforms/Scalar/ConstraintElimination.cpp]: Done.

**Review by nikic — APPROVED** [review]
LGTM

**Review by fhahn — APPROVED** [review]
LGTM, thanks!

**Pull request merged**